### DISCUSSION TITLE
How to dynamically restart Application

### DISCUSSION THREAD

**Discussion by Juff-Ma** [AvaloniaUI/Avalonia#15363] [Q&A] [answered]
How to restart an App dynamically? I mean stopping the application completetly and starting it again, only closing the window isn't possible as the app will also run in framebuffer mode.
If I try to just start the app builder again it throws an exception with "Setup was already called on one of AppBuilder instances.". In older avalonia verions (until 0.10.22 it works, maybe also early 11.0 versions but in this version definitely works) you could circumvent this by using SetupWithoutStarting like so:

```
using System;

using Avalonia;
using Avalonia.Controls.ApplicationLifetimes;

namespace RestartTest
{
    public class Program
    {
        static void Main(string[] args)
        {
            var builder =
                AppBuilder.Configure<App>()
                    .UsePlatformDetect()
                    .LogToTrace()
                    .SetupWithoutStarting();

            while (true)
            {
                var lifetime = new ClassicDesktopStyleApplicationLifetime();
                builder.Instance.ApplicationLifetime = lifetime;
                builder.Instance.OnFrameworkInitializationCompleted();
                lifetime.Start(args);

                ((ClassicDesktopStyleApplicationLifetime)Application.Current.ApplicationLifetime).Dispose();
            }
        }
    }
}
```

However in Avalonia 11 the same code throws System.InvalidOperationException: "It's not possible to change ApplicationLifetime after Application was initialized.". 

Is there a way to circumvent this? It is requirement for my app to dynamically close itself and i didn't find a way to hide a framebuffer app without closing it.

All ways to do this were done in Avalonia 0.10 where but i would rather stay on 11.

**Comment by maxkatz6** [comment]
@Juff-Ma you can do something like this, without lifetimes at all:
```
AppBuilder.Configure<App>().UsePlatformDetect().SetupWithoutStarting();

while (true)
{
   new MainWindow().Show();
   Dispatcher.PushFrame(new DispatcherFrame());
}
```

**Reply by maxkatz6** [reply]
But it's not really a restart of the app. Just restarting Avalonia main loop. App instance is not recreated either.

**Reply by Juff-Ma** [reply]
> @Juff-Ma you can do something like this, without lifetimes at all:
> ```
> AppBuilder.Configure<App>().UsePlatformDetect().SetupWithoutStarting();
> 
> while (true)
> {
>    new MainWindow().Show();
>    Dispatcher.PushFrame(new DispatcherFrame());
> }
> ```

That isn't possible in framebuffer mode as far as I know and framebuffer mode is a requirement for me.

**Reply by maxkatz6** [reply]
I don't think you can restart StartLinuxFbDev then. It will be easier to figure out how to reopen window or root view.

**Reply by Juff-Ma** [reply]
> I don't think you can restart StartLinuxFbDev then. It will be easier to figure out how to reopen window or root view. 

I wouldn't need to restart it but restarting it was the only way I found to get it to stop rendering. I need to start a video player using DRM, wait for it to close and then restart the avalonia GUI, the app is written for embedded, desktop is only used for testing.

On the other hand, why was this introduced? Is there a significant change in avalonia 11 that requires this change or can the avalonia 10 behavior somehow be replicated?

**Reply by maxkatz6** [reply]
> I wouldn't need to restart it but restarting it was the only way I found to get it to stop rendering. I need to start a video player using DRM, wait for it to close and then restart the avalonia GUI, the app is written for embedded, desktop is only used for testing.

You can call protected StartRendering and StopRendering methods from Window. But not sure if it will work as you want. 

> On the other hand, why was this introduced? Is there a significant change in avalonia 11 that requires this change or can the avalonia 10 behavior somehow be replicated?

What exactly? Not possibility to change ApplicationLifetime after app already started? Because it was a logical bug, and didn't make sense. By recreating application lifetime you **do not** restart the app.

**Reply by Juff-Ma** [reply]
> You can call protected StartRendering and StopRendering methods from Window. But not sure if it will work as you want. 

That's the problem, I don't have a window, it's a single view embedded app.

> What exactly? Not possibility to change ApplicationLifetime after app already started? Because it was a logical bug, and didn't make sense. By recreating application lifetime you **do not** restart the app.

Even if it was technically a bug, removing it and not providing an alternative is actively removing functionality.

**Reply by maxkatz6** [reply]
> That's the problem, I don't have a window, it's a single view embedded app.

If you set IsVisible to false, it will disable any rendering of the subtree.

**Reply by Juff-Ma** [reply]
> > That's the problem, I don't have a window, it's a single view embedded app.
> 
> If you set IsVisible to false, it will disable any rendering of the subtree.

Would this free up `/dev/dri/card1` for use of another program?

**Reply by maxkatz6** [reply]
No. Restarting lifetime also wouldn't.

**Reply by Juff-Ma** [reply]
> No. Restarting lifetime also wouldn't. 

Why not? Why wouldn't telling the app to exit and restarting it also stop it from rendering and then start rendering again? Why would avalonia capture a rendering device if it isn't currently using it (for example the application being stopped)? I mean avalonia doesn't block the drm device if you stop the app using CTR+C. (not to be rude but I honestly don't understand why avalonia would do this)

And if this wouldn't work how would one do this? It is a requirement for our app.

**Reply by MikeCodesDotNET** [reply]
Given the nature and complexity of the question, I recommend you contact us to arrange commercial support. 

We offer various [support agreements](https://avaloniaui.net/) to provide in-depth support from the Avalonia core team.

**Reply by Juff-Ma** [reply]
Sadly we are a small group working in an educational institution and our budget is too limited to get paid support.

**Comment by Pratham95VP** [comment]
` string exePath = Process.GetCurrentProcess().MainModule.FileName;
 Process.Start(new ProcessStartInfo(exePath) { UseShellExecute = true });
 Environment.Exit(0);`  try this.

**Reply by Juff-Ma** [reply]
That would be the nuclear option, while it would work I don't like it and would like to consider all other options first.

**Reply by damonpkuml** [reply]
The restart does not take effect on Mac.

**Reply by Juff-Ma** [reply, marked as answer]
> The restart does not take effect on Mac.

OP here. In the end after I stopped and actually thought about what I wanted, Avalonia wasn't actually the place to go for this. It's a nice UI framework but it isn't its or its developers job to realize complex lifecycle management.

In the end the solution for me was to split lifecycle/business and UI into 2 separate applications that communicated via IPC. That way I could keep using Avalonia in a way it was actually designed for and still use a complex lifecycle by just making the UI app a subprocess of the business app.

**Reply by damonpkuml** [reply]
on Window:
 ```
var exePath = Environment.ProcessPath!;
 Process.Start(new ProcessStartInfo
 {
     FileName = exePath,
     UseShellExecute = true
 });
```
 on Mac:
 ```
 var appBundlePath = Environment.ProcessPath!;

 Process.Start(new ProcessStartInfo
 {
     FileName = "open",
     Arguments = $"-n \"{appBundlePath}\"",
     UseShellExecute = false
 });
```
 
 Avalonia is a very good framework. Developers can handle platform-specific differences and achieve consistent implementations across different platforms.
